File: week4/p9.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def getIntersectionNode(head1, head2):
    p1, p2 = head1, head2
    while p1 != p2:
        p1 = p1.next if p1 else head2
        p2 = p2.next if p2 else head1
    return p1

if __name__ == "__main__":
    common = Node(8)
    common.next = Node(4)
    common.next.next = Node(5)

    head1 = Node(4)
    head1.next = Node(1)
    head1.next.next = common

    head2 = Node(5)
    head2.next = Node(6)
    head2.next.next = Node(1)
    head2.next.next.next = common

    print(getIntersectionNode(head1, head2).data)  # 8

File: week4/p9.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node* getIntersectionNode(struct Node* head1, struct Node* head2) {
    struct Node* p1 = head1;
    struct Node* p2 = head2;
    while (p1 != p2) {
        p1 = p1 ? p1->next : head2;
        p2 = p2 ? p2->next : head1;
    }
    return p1;
}

int main() {
    struct Node* common = newNode(8);
    common->next = newNode(4);
    common->next->next = newNode(5);

    struct Node* head1 = newNode(4);
    head1->next = newNode(1);
    head1->next->next = common;

    struct Node* head2 = newNode(5);
    head2->next = newNode(6);
    head2->next->next = newNode(1);
    head2->next->next->next = common;

    struct Node* intersect = getIntersectionNode(head1, head2);
    if (intersect) printf("%d\n", intersect->data);
    return 0;
}
